drgn_tools/task.py

```diff
@@ -185,6 +195,173 @@ def get_command(task: Object) -> str:
+    should be used for single task
+    """
+    MM_FILEPAGES = prog.constant("MM_FILEPAGES").value_()
+    MM_ANONPAGES = prog.constant("MM_ANONPAGES").value_()
+    try:
+        MM_SHMEMPAGES = prog.constant("MM_SHMEMPAGES").value_()
+    except LookupError:
+        MM_SHMEMPAGES = -1
+
+    try:
+        filerss = task.mm.rss_stat.count[MM_FILEPAGES].counter.value_()
+        anonrss = task.mm.rss_stat.count[MM_ANONPAGES].counter.value_()
+        if MM_SHMEMPAGES > 0:
+            shmrss = task.mm.rss_stat.count[MM_SHMEMPAGES].counter.value_()
+        else:
+            shmrss = 0
+
+        rss_mm = filerss + anonrss + shmrss
+
+    except drgn.FaultError:
+        rss_mm = 0
+
+    gid = task.tgid.value_()
+
+    rss_task = rss_task_dict.get(gid)
+
+    if gid in rss_task_dict:
+        rss_task = rss_task_dict[gid]
+    else:
+        rss_task = 0
+        ltask = task.group_leader
+
+        task_filerss = ltask.rss_stat.count[MM_FILEPAGES].value_()
+        task_anonrss = ltask.rss_stat.count[MM_ANONPAGES].value_()
+        if MM_SHMEMPAGES > 0:
+            task_shmrss = ltask.rss_stat.count[MM_SHMEMPAGES].value_()
+        else:
+            task_shmrss = 0
+
+        rss_task += task_filerss + task_anonrss + task_shmrss
+
+        for gtask in list_for_each_entry(
+            "struct task_struct",
+            task.group_leader.thread_group.address_of_(),
+            "thread_group",
+        ):
+            task_filerss = gtask.rss_stat.count[MM_FILEPAGES].value_()
+            task_anonrss = gtask.rss_stat.count[MM_ANONPAGES].value_()
+            if MM_SHMEMPAGES > 0:
+                task_shmrss = gtask.rss_stat.count[MM_SHMEMPAGES].value_()
+            else:
+                task_shmrss = 0
+
+            rss_task += task_filerss + task_anonrss + task_shmrss
+
+        rss_task_dict[gid] = rss_task
+
+    rss = rss_mm + rss_task
+    page_size = prog["PAGE_SIZE"].value_()
+    rss_kb = (rss * page_size) // ByteToKB
+
+    return rss_kb
+
+
+def get_pct_mem(prog: Program, task: Object) -> str:
+    """
+    :returns %MEM value
+    """
+    gid = task.tgid.value_()
+    rss_task = rss_task_dict_global[gid]
+    rss_mm = rss_mm_dict_global[task.pid.value_()]
+    tot_rss = rss_mm + rss_task
+
+    page_shift = prog["PAGE_SHIFT"].value_()
+    high_mem = prog["high_memory"].value_()
+
+    if prog.platform.arch == Architecture.X86_64:
+        try:
+            page_offset_base = prog["page_offset_base"].value_()
+        except KeyError:
+            # Prior to 021182e52fe0 ("x86/mm: Enable KASLR for physical mapping
+            # memory regions"), page_offset_base did not exist and there was
+            # just a constant for the beginning of the memory map.
+            page_offset_base = 0xFFFF880000000000
+
+        total_pages = (high_mem - page_offset_base) >> page_shift
+    else:
+        raise NotImplementedError(
+            "%MEM calculation is not (yet) implemented for aarch64"
+        )
```

Comment:
What if we used the function `totalram_pages()` instead of this architecture specific computation of `total_pages`?

This may not be exactly the same, but it strikes me that `high_mem` is a pretty archaic remnant of 32-bit kernels at this point.

Reply:
I had tested with `totalram_pages()` initially but as you said the value will not be exactly same, it will be bit different like below:

total_pages = 4456448 ( with above implementation)
```
4124985  4114326  0    0xffff91b3e199b000  S   1.6   1711276  290144  osms-agent
4127087  4114326  0    0xffff91b5b3161800  S   1.6   1711276  290144  osms-agent
4129199  4114326  2    0xffff91b430f6c800  S   1.6   1711276  290144  osms-agent
4131335  4114326  0    0xffff91b66f92e000  S   1.6   1711276  290144  osms-agent
```

totalram_pages = 4022584
```
4124985  4114326  0    0xffff91b3e199b000  S   1.8   1711276  290144  osms-agent
4127087  4114326  0    0xffff91b5b3161800  S   1.8   1711276  290144  osms-agent
4129199  4114326  2    0xffff91b430f6c800  S   1.8   1711276  290144  osms-agent
4131335  4114326  0    0xffff91b66f92e000  S   1.8   1711276  290144  osms-agent
```

If this much difference in the output is fine then we can use totalram_pages() function. Please let me know your thought on this.

Reply:
I took a second look and still think `totalram_pages()` is the best way. I skimmed the code from a couple locations to ensure I'm correct.

1. The kernel uses the same logic as `totalram_pages()` to report `MemTotal` in `/proc/meminfo`.
2. The userspace `ps(1)` utility (as implemented by procps-ng) divides the process RSS by the `MemTotal` value to get `%MEM`.
3. Crash has a more convoluted path toward getting the total memory, I didn't completely follow it during my brief skim.

As far as I'm concerned, `ps(1)` is the ground truth here, not crash. So let's go with `totalram_pages()` and accept the difference from crash!

Reply:
okay, I have made changes in latest code. Thanks!